tests/pstdarg_test.cpp

```diff
@@ -0,0 +1,184 @@
+	P_DEBUG ("Unstacking a " #type);			\
+	P_TEST_CHECK (type##_va_arg(ap) == wantedvalue);	\
+  } while (0)
+
+#define P_TEST_VA_ARG_HEAD(ap) do {				\
+	P_TEST_VA_ARG(ap, pint8, pint8_var[0]);			\
+	P_TEST_VA_ARG(ap, puint8, puint8_var[0]);		\
+	P_TEST_VA_ARG(ap, pint16, pint16_var[0]);		\
+	P_TEST_VA_ARG(ap, puint16, puint16_var[0]);		\
+	P_TEST_VA_ARG(ap, pint32, pint32_var[0]);		\
+	P_TEST_VA_ARG(ap, puint32, puint32_var[0]);		\
+	P_TEST_VA_ARG(ap, pint64, pint64_var[0]);		\
+	P_TEST_VA_ARG(ap, puint64, puint64_var[0]);		\
+	P_TEST_VA_ARG(ap, pshort, pshort_var[0]);		\
+	P_TEST_VA_ARG(ap, plong, plong_var[0]);			\
+	P_TEST_VA_ARG(ap, pchar, pchar_var[0]);			\
+	P_TEST_VA_ARG(ap, ppointer, ppointer_var[0]);		\
+	P_TEST_VA_ARG(ap, pfloat, pfloat_var[0]);		\
+	P_TEST_VA_ARG(ap, pdouble, pdouble_var[0]);		\
+  } while (0)
+
+#define P_TEST_VA_ARG_TRAIL(ap) do {				\
+	P_TEST_VA_ARG(ap, pdouble, pdouble_var[0]);		\
+	P_TEST_VA_ARG(ap, pfloat, pfloat_var[0]);		\
+	P_TEST_VA_ARG(ap, ppointer, ppointer_var[0]);		\
+	P_TEST_VA_ARG(ap, pchar, pchar_var[0]);			\
+	P_TEST_VA_ARG(ap, plong, plong_var[0]);			\
+	P_TEST_VA_ARG(ap, pshort, pshort_var[0]);		\
+	P_TEST_VA_ARG(ap, puint64, puint64_var[0]);		\
+	P_TEST_VA_ARG(ap, pint64, pint64_var[0]);		\
+	P_TEST_VA_ARG(ap, puint32, puint32_var[0]);		\
+	P_TEST_VA_ARG(ap, pint32, pint32_var[0]);		\
+	P_TEST_VA_ARG(ap, puint16, puint16_var[0]);		\
+	P_TEST_VA_ARG(ap, pint16, pint16_var[0]);		\
+	P_TEST_VA_ARG(ap, puint8, puint8_var[0]);		\
+	P_TEST_VA_ARG(ap, pint8, pint8_var[0]);			\
+  } while (0)
+
+static void variadic_function_copy_all (pint unused, p_va_list ap)
+{
+	P_UNUSED (unused);
+	P_DEBUG ("Unstacking a copy of all the arguments");
+	P_TEST_VA_ARG_HEAD (ap);
+	P_TEST_VA_ARG_TRAIL (ap);
+}
+
+static void variadic_function_copy_trail (pint unused, p_va_list ap)
+{
+	P_UNUSED (unused);
+	P_DEBUG ("Unstacking second part of the arguments");
+	P_TEST_VA_ARG_TRAIL (ap);
+}
+
+static void variadic_function (pint unused, ...)
+{
+	p_va_list ap;
+	p_va_list ap2;
+
+	p_va_start (ap, unused);
+
+	P_DEBUG ("Copy of arguments");
+	p_va_copy (ap2, ap);
+	variadic_function_copy_all (unused, ap2);
+	p_va_end (ap2);
+
+	P_DEBUG ("Unstacking first part of arguments");
+	P_TEST_VA_ARG_HEAD (ap);
+
+	P_DEBUG ("Copy of arguments at current unstack state, i.e. in the middle");
+	p_va_copy (ap2, ap);
+	variadic_function_copy_trail (unused, ap2);
+	p_va_end (ap2);
+
+	P_DEBUG ("Unstacking second part of arguments");
+	P_TEST_VA_ARG_TRAIL (ap);
+
+	p_va_end (ap);
+}
+
+P_TEST_CASE_BEGIN (pstdarg_general_test)
+{
+	p_libsys_init ();
+
+	pint32_var[0] = P_MININT16;
+	pint32_var[0] <<= 16;
+	pint32_var[1] = P_MAXINT16;
+	pint32_var[1] <<= 16;
+	puint32_var[1] = P_MAXUINT16;
+	puint32_var[1] <<= 16;
+
+	pint64_var[0] = pint32_var[0];
+	pint64_var[0] <<= 32;
+	pint64_var[1] = pint32_var[1];
+	pint64_var[1] <<= 32;
+	puint64_var[1] = puint32_var[1];
+	puint64_var[1] <<= 32;
+
+	variadic_function (0,
```

Comment:
Why do we use only the first elements in arrays *_var, but define 2 items in each? Should we also use the second ones?

Reply:
Yep - should be [1] in the trail macro and in the second part of the push your're right - will fix that later.
Thanks for spotting this -;...

Reply:
Fixed with aaaad4d58cbdb973fe9710a3ed3b5c182cc4f7f0